chacha20/src/nivc.nr

```diff
@@ -0,0 +1,334 @@
+                ciphertext[block_idx * 16 + j] = key_stream[j] ^ plaintext_bits[block_idx * 16 + j];
+            }
+            counter = add32(counter, 1);
+        }
+
+        // Step 2: Convert ciphertext to big endian and handle padding
+        for i in 0..(DATA_BYTES / 4) {
+            // input is words(32bits)
+            let be_word = little_to_big_endian(ciphertext[i]);
+
+            for j in 0..4 {
+                let byte_idx = i * 4 + j;
+                let is_padding_current = is_padding[byte_idx];
+
+                let byte = if is_padding_current {
+                    0
+                } else {
+                    get_byte_from_word(be_word, j as u8)
+                };
+                big_endian_ciphertext[byte_idx] = byte;
+            }
+        }
+
+        // Step 3: Calculate digests
+        let mut part_ciphertext_digest = Field::default();
+        let mut plaintext_digest = Field::default();
+
+        part_ciphertext_digest = DataHasher::new(big_endian_ciphertext, step_in[10]);
+        plaintext_digest = polynomial_digest(
+            zeroed_plaintext.map(|x| x as Field),
+            self.ciphertext_digest,
+            step_in[1],
+        );
+
+        for i in 0..DATA_BYTES {
+            if (is_padding[i]) {
+                // do nothing
+            } else {
+                ciphertext_digest_pow *= self.ciphertext_digest;
+            }
+        }
+
+        // Step 4: Update step_out
+        step_out[0] = step_in[0] + step_in[10] - part_ciphertext_digest + plaintext_digest;
+        step_out[1] = ciphertext_digest_pow;
+        step_out[2] = step_in[2]; // Ciphertext digest POW accumulator
+        step_out[3] = polynomial_digest([1, 0, 0, 0, 0, 0, 0, 1], self.ciphertext_digest, 1); // Default machine state digest
+        step_out[10] = part_ciphertext_digest;
+
+        step_out[4] = step_in[4];
+        step_out[5] = step_in[5];
+        step_out[6] = 0; // Body ciphertext digest pow counter
+        step_out[7] = step_in[7];
+        step_out[8] = step_in[8];
+        step_out[9] = step_in[9];
+
+        step_out
+
+        // step_out
+    }
+}
+
+// TODO: can we use generics for input?
+pub fn polynomial_digest<let N: u32>(
+    input: [Field; N],
+    digest_ratio: Field,
+    default_multiplier: Field,
+) -> Field {
+    let mut digest = Field::default();
+
+    let mut multiplier = default_multiplier;
+    for item in input {
+        digest += multiplier * item; // TODO: check if subtyping is expensive
+        multiplier *= digest_ratio;
+    }
+
+    digest
+}
+
+use std::hash::{self, poseidon::bn254::hash_2};
+
+pub struct DataHasher<let DATA_BYTES: u32> {}
+
+impl<let DATA_BYTES: u32> DataHasher<DATA_BYTES> {
+    fn new(data: [u8; DATA_BYTES], seed: Field) -> Field {
+        // ceil(DATA_BYTES/16) == (DATA_BYTES + 15) / 16 + 1
+        let mut hashes: [Field; (DATA_BYTES + 15) / 16 + 1] =
+            [Field::default(); (DATA_BYTES + 15) / 16 + 1];
+        hashes[0] = seed;
```

Comment:
An `hashes` array can be a Field if you do not have to record all previous hashes.
```suggestion
        let mut hash: Field = Field::default();
        hash = seed;
```

Reply:
okay